unstable test TestRenameTableIntermediateState

## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

<!-- a step by step guide for reproducing the bug. -->

### 2. What did you expect to see? (Required)

### 3. What did you see instead (Required)

```
     table_test.go:560: 
        	Error Trace:	ddl/table_test.go:560
        	            				ddl/util/callback/callback.go:124
        	            				ddl/job_table.go:449
        	            				util/wait_group_wrapper.go:154
        	            				src/runtime/asm_amd64.s:1598
        	Error:      	Expected value not to be nil.
        	Test:       	TestRenameTableIntermediateState
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x839a89f]
goroutine 573593 [running]:
github.com/pingcap/tidb/ddl_test.TestRenameTableIntermediateState.func1(0xc017fce840?)
	ddl/table_test.go:561 +0xff
github.com/pingcap/tidb/ddl/util/callback.(*TestDDLCallback).OnJobUpdated(0xc01e561b20, 0xc017fce840?)
	ddl/util/callback/callback.go:124 +0x1ea
github.com/pingcap/tidb/ddl.(*ddl).delivery2worker.func1()
	ddl/job_table.go:449 +0xdef
github.com/pingcap/tidb/util.(*WaitGroupWrapper).Run.func1()
	util/wait_group_wrapper.go:154 +0x6b
created by github.com/pingcap/tidb/util.(*WaitGroupWrapper).Run
	util/wait_group_wrapper.go:152 +0x8d
-- 
```

https://tiprow.hawkingrei.com/view/gs/pingcapprow/pr-logs/pull/pingcap_tidb/44947/tiprow_fast_test/1673232837370712064
### 4. What is your TiDB version? (Required)

<!-- Paste the output of SELECT tidb_version() -->

--- a/pkg/ddl/table_test.go
+++ b/pkg/ddl/table_test.go
@@ -549,20 +549,31 @@ func TestRenameTableIntermediateState(t *testing.T) {
 		{"rename table db2.t to db1.t;", "insert into db1.t values(1);", "", "db1.t"},
 	}
 
+	var finishedJobID int64
 	for _, tc := range testCases {
 		hook := &callback.TestDDLCallback{Do: dom}
 		runInsert := false
+		var jobID int64 = 0
 		fn := func(job *model.Job) {
+			if job.ID <= finishedJobID {
+				// The job has been done, OnJobUpdated may be invoked later asynchronously.
+				// We should skip the done job.
+				return
+			}
 			if job.Type == model.ActionRenameTable &&
 				job.SchemaState == model.StatePublic && !runInsert && !t.Failed() {
 				_, err := tk2.Exec(tc.insertSQL)
+				// In rename table intermediate state, new table is public.
 				if len(tc.errMsg) > 0 {
+					// Old table should not be visible to DML.
 					assert.NotNil(t, err)
 					assert.Equal(t, tc.errMsg, err.Error())
 				} else {
+					// New table should be visible to DML.
 					assert.NoError(t, err)
 				}
 				runInsert = true
+				jobID = job.ID
 			}
 		}
 		hook.OnJobUpdatedExported.Store(&fn)
@@ -575,6 +586,7 @@ func TestRenameTableIntermediateState(t *testing.T) {
 			result.Check(testkit.Rows("1"))
 		}
 		tk.MustExec(fmt.Sprintf("delete from %s;", tc.finalDB))
+		finishedJobID = jobID
 	}
 	dom.DDL().SetHook(originHook)
 }

